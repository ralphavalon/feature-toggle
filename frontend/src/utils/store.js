import React, { useState } from 'react';

export const StoreContext = React.createContext(null);

export default ({ children }) => {
  const [features, setFeatures] = useState([]);
  const [hasError, setHasError] = useState(false);

  const store = {
    features: [features, setFeatures],
    error: [hasError, setHasError]
  };
  return (
    <StoreContext.Provider value={store}>
      {children}
    </StoreContext.Provider>
  );
};

