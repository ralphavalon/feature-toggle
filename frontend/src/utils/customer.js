import React, { useContext } from 'react';
import { StoreContext } from './store';

const getCustomerIds = () => {
  const { features: [features, setFeatures] } = useContext(StoreContext);

  let customerIds = [];
  features.forEach(f => {
    customerIds = customerIds.concat(f.customerIds.filter((item) => customerIds.indexOf(item) < 0));
  });
  return customerIds;
};

export { getCustomerIds };
