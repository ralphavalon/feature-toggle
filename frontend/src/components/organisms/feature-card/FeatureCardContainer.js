import React, { useState, useContext, useEffect } from 'react';
import axios from 'axios';
import serialize from 'form-serialize';
import PropTypes from 'prop-types';

import FeatureCard from './FeatureCard';
import { StoreContext } from '../../../utils/store';

const FeatureCardContainer = props => {
  const { features: [features, setFeatures] } = useContext(StoreContext);
  const [isLoading, setIsLoading] = useState(false);
  const [hasError, setHasError] = useState(false);

  const isNew = () => !props.id;

  const createOrUpdateFeature = async feature => {
    try{
      setIsLoading(true);
      const result = await axios.put(`${process.env.REACT_APP_FEATURE_TOGGLE_URL}`, feature);
      // feature.id = result.data.id;
      if(isNew()) {
        setFeatures([...features, { ...feature, id: result.data.id}]);
      }
    } catch (error) {
      setHasError(true);
      console.error(error);
    } finally {
      setIsLoading(false);
    }
  };

  const onSubmit = async event => {
    event.preventDefault();

    const feature = serialize(event.target, { hash: true });

    if(!Array.isArray(feature.customerIds) && feature.customerIds) {
      feature.customerIds = [feature.customerIds];
    }
    if(feature.expiresOn) {
      feature.expiresOn = feature.expiresOn.replace(/ /g,'');
    }

    await createOrUpdateFeature(feature);

    if(props.onSubmit) {
      props.onSubmit();
    }
  };

  return (
    <FeatureCard {...props} isNew={isNew()} onSubmit={onSubmit}/>
  );
};

FeatureCardContainer.propTypes = {
  onSubmit: PropTypes.func,
  id: PropTypes.string,
};

export default FeatureCardContainer;
