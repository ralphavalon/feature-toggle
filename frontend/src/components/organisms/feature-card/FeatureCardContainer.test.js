import React from 'react';
import axios from 'axios';
import { shallow, mount } from 'enzyme';
import serialize from 'form-serialize';
import { act } from 'react-dom/test-utils';

import * as Customer from '../../../utils/customer';
import FeatureCardContainer from './FeatureCardContainer';
import { StoreContext } from '../../../utils/store';

jest.mock('axios');
jest.mock('form-serialize');
jest.mock('../../../utils/customer');

const mockGetCustomerIds = value => {
  Customer.getCustomerIds.mockImplementation(() => value);
}
let event = {
  preventDefault: jest.fn(),
  target: {
    displayName: 'displayName',
    technicalName: 'technicalName',
    active: true,
    inverted: true,
    customerIds: ['123', '321'],
    expiresOn: '2021-05-05T 11:22:33'
  }
};

const expectedRequest = {
  ...event.target,
  expiresOn: '2021-05-05T11:22:33',
  id: 'id'
}

beforeEach(() => {
  mockGetCustomerIds([]);
  event = {
    preventDefault: jest.fn(),
    target: {
      displayName: 'displayName',
      technicalName: 'technicalName',
      active: true,
      inverted: true,
      customerIds: ['123', '321'],
      expiresOn: '2021-05-05T 11:22:33'
    }
  };
});

afterEach(() => {
  jest.clearAllMocks();
});

const featureCardContainerOptionalProps = {
  onSubmit: jest.fn(),
  id: 'id'
};


const runInAct = async (wrapper) => {
  await act(async () => {
    await new Promise(resolve => setTimeout(resolve, 0));
    wrapper();
  });
};

describe('FeatureCardContainer', () => {
  it('renders correctly', () => {
    let wrapper = shallow(
      <StoreContext.Provider value={{
        features: [[], jest.fn()]
      }}>
        <FeatureCardContainer {...featureCardContainerOptionalProps} />
      </StoreContext.Provider>
    );

    expect(wrapper).toMatchSnapshot();

    wrapper = shallow(
      <StoreContext.Provider value={{
        features: [[], jest.fn()]
      }}>
        <FeatureCardContainer {...{...featureCardContainerOptionalProps, id: null }} />
      </StoreContext.Provider>
    );
  });

  test.each([
    {
      props: {...featureCardContainerOptionalProps, id: null }, 
      expectedRequest: { 
        ...expectedRequest,
        id: undefined
      },
      serializeMock: event.target
    },
    {
      props: featureCardContainerOptionalProps, 
      expectedRequest,
      serializeMock: {...event.target, id: featureCardContainerOptionalProps.id }
    },
  ])(`should create/update feature when onSubmit`, async ({ props, expectedRequest, serializeMock }) => {
    axios.put.mockResolvedValue({ data: { id: featureCardContainerOptionalProps.id } });
    serialize.mockImplementation(() => serializeMock);

    const wrapper = mount(
      <StoreContext.Provider value={{
        features: [[], jest.fn()]
      }}>
        <FeatureCardContainer {...props} />
      </StoreContext.Provider>
    );

    await runInAct(() => wrapper.find('FeatureCard').prop('onSubmit')(event));

    expect(wrapper.find('FeatureCardContainer').prop('onSubmit')).toEqual(featureCardContainerOptionalProps.onSubmit);
    expect(event.preventDefault).toHaveBeenCalledTimes(1);
    expect(serialize).toHaveBeenCalledTimes(1);
    expect(featureCardContainerOptionalProps.onSubmit).toHaveBeenCalledTimes(1);
    expect(axios.put).toHaveBeenCalledWith(expect.anything(), expectedRequest);
  });
})

