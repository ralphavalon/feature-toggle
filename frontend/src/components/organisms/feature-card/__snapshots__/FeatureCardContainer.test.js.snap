// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatureCardContainer renders correctly 1`] = `
<FeatureCardContainer
  id="id"
  onSubmit={[MockFunction]}
/>
`;
